discord/voice_client.py

```diff
@@ -643,3 +691,163 @@ def send_audio_packet(self, data, *, encode=True):
+
+        Raises
+        ------
+        ClientException
+            Not connected to a voice channel.
+        ClientException
+            Already recording.
+        ClientException
+            Must provide a Sink object.
+        """
+        if not self.is_connected():
+            raise ClientException('Not connected to voice channel.')
+        if self.recording:
+            raise ClientException("Already recording.")
+        if not isinstance(sink, Sink):
+            raise ClientException("Must provide a Sink object.")
+
+        self.empty_socket()
+
+        self.decoder = opus.DecodeManager(self)
+        self.decoder.start()
+        self.recording = True
+        self.sink = sink
+        sink.init(self)
+
+        t = threading.Thread(target=self.recv_audio, args=(sink, callback, *args,))
+        t.start()
+
+    def stop_recording(self):
+        """Stops the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+        """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.decoder.stop()
+        self.recording = False
+        self.paused = False
+
+    def pause_recording(self):
+        """Pauses or unpauses the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+         """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.paused = {True: False, False: True}[self.paused]
+
+    def empty_socket(self):
+        while True:
+            ready, _, _ = select.select([self.socket], [], [], 0.0)
+            if len(ready) == 0:
+                break
+            for s in ready:
+                s.recv(4096)
+
+    def recv_audio(self, sink, callback, *args):
+        #  Gets data from _recv_audio and sorts
+        #  it by user, handles pcm files and
+        #  silence that should be added.
+
+        self.user_timestamps = {}
+        self.starting_time = time.perf_counter()
+        while self.recording:
+            ready, _, err = select.select([self.socket], [],
+                                          [self.socket], 0.01)
+            if not ready:
+                if err:
+                    print("Socket error")
+                continue
+
+            try:
+                data = self.socket.recv(4096)
+            except OSError:
+                self.stop_recording()
+                continue
+
+
+            self.unpack_audio(data)
+
+        self.stopping_time = time.perf_counter()
+        self.sink.cleanup()
+        try:
+            callback = asyncio.run_coroutine_threadsafe(callback(self.sink, *args), self.loop)
+            result = callback.result()
+        except Exception as exc:
+            raise exc
+        else:
+            if result is not None:
+                print(result)
```

Comment:
So that I can run it from start_recording which is not a coroutine because I didn't want it to be a coroutine just how vc.play is not a coroutine. The only problem with that is running the callback function, which I used asyncio.run_coroutine_threadsafe for. If it becomes/is a problem then I can make it a coroutine function.

Reply:
Oh I see now, you call this in a thread. Yeah I am not sure